### ISSUE TITLE
Working directory

### ISSUE THREAD

**Issue by TomdeGraaf** [rationalmatter/Juno-Issue-Tracker#181]
**Describe the bug**
Often, but not always, when I open an notebook file from a folder, and call on for instance other files that live in the same folder, the files are not found. Then when I change working dir to the same folder the notebook is in (which I think already WAS the working dir when I open the notebook there), everything works and files are found.

**To Reproduce**
2. Tap on '....'

**Expected behavior**
I expected that opening a notebook in a location automatically sets the current working dir to be that same folder that notebook lives in. This does not systematically happens for me.

**Screenshots**
If applicable, add screenshots or a screen recording to help explain your problem.

**Environment (please complete the following information):**
 - Device: [e.g. iPad Pro 11"] iPad Pro 2018 12.9 inch
 - iOS version: [e.g. iOS 12.2] iOS 13.6
 - App version [e.g. 1.0.0] 1.2.6

**Comment by navoshta** [comment]
Thanks for reporting @TomdeGraaf — you are right that Juno should switch to notebook's directory as cwd automatically. To be precise, the logic is as follows:

1. If notebooks is opened for the first time, and is in a directory which Juno has access to, notebook's directory becomes notebook's cwd.
2. If notebooks is opened for the first time, and Juno _doesn't_ have access to the directory (say, you opened a notebook in another app's sandbox), Juno will set cwd to a temporary folder and display a "locked folder" icon in the top bar.
3. If you have opened this notebook before and switched notebook's cwd previously (either by granting access to notebook's directory, or by selecting any other location), Juno will attempt to restore the same cwd the next time you open that notebook. If it fails to do so (say, directory was moved, or iOS revoked Juno's access to it), it will fallback to temporary folder as cwd, and will display a "locked folder" icon.

Is there any consistency to which notebooks _don't_ set their cwd correctly? Any chance you had changed cwd for those notebooks previously to something else? Or perhaps you are opening a notebook in another app's directory?

**Comment by craldaz** [comment]
Hi @navoshta I'm a new user to Juno, I'm working on an iPad Pro 12.9 Inch 2020. Just to be clear -- Juno should be able to run from other directories like Dropbox, ICloud, etc?

Because if that is so then I can't get it to work. When I type os.getcwd() in a notebook running on Dropbox or anywhere else, it always give /private/var/mobile/Containers/Data/Application/FB.../Caches

What is the meaning of your last sentence "Or perhaps you are opening a notebook in another app's directory?" -- Isn't that the point, to open notebooks in other directories like Dropbox, iCloud, etc.?

**Comment by navoshta** [comment]
> Juno should be able to run from other directories like Dropbox, ICloud, etc?

Yes, but you need to explicitly grant access to those directories (or anything outside of app's sandbox) by tapping the "lock" button in the top bar. Also, some cloud storage providers will not support this — e.g. Dropbox, although iCloud should work fine.

> When I type os.getcwd() in a notebook running on Dropbox or anywhere else, it always give /private/var/mobile/Containers/Data/Application/FB.../Caches

Yep, this is means Juno couldn't get access to notebook's directory, and had to set your cwd to a fallback value (`/Temp`).

> What is the meaning of your last sentence "Or perhaps you are opening a notebook in another app's directory?" -- Isn't that the point, to open notebooks in other directories like Dropbox, iCloud, etc.?

Yes, opening notebooks in another app's directory, in iCloud or Dropbox is essentially the same thing, from the file system access standpoint. To Juno, this means opening a notebook outside of app's sandbox, and due to how iOS file system works, Juno will be able to get access to the notebook itself, and save all changes to the original location. 

The _directory_ access is a whole different beast though, and OS does _not_ automatically grant apps access to directories if they happen to be editing a file in it. For that, you need to explicitly grant access _to the directory_: either tap the "lock" button in the top bar, or better go to notebook menu (tap notebook name in the top bar) and select "Change working directory..." option. This will let you (a) choose a new cwd using a system dialogue, and (b) will automatically request access to it from the OS.

**Comment by alexstaravoitau** [comment]
Closing this ticket for now, but please feel free to reopen (or submit a new one), if this remains a problem. There is also a doc on the topic now: https://juno.sh/ios-file-system/
